pkg/controller/metal3remediation/metal3remediation_controller.go

```diff
@@ -0,0 +1,470 @@
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	infrav1 "github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	"github.com/openshift/cluster-api-provider-baremetal/pkg/baremetal"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Metal3RemediationReconciler reconciles a Metal3Remediation object.
+type Metal3RemediationReconciler struct {
+	client.Client
+	ManagerFactory baremetal.ManagerFactoryInterface
+	Log            logr.Logger
+}
+
+// Generate RBAC
+//go:generate go run ../../../vendor/sigs.k8s.io/controller-tools/cmd/controller-gen paths=./... rbac:roleName=machine-api-controllers-baremetal crd output:dir:=./../../../config/rbac
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3remediations,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3remediations/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=machine.openshift.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;update;patch;delete
+
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts/status,verbs=get;update;patch
+
+// Reconcile handles Metal3Remediation events.
+func (r *Metal3RemediationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
+	remediationLog := r.Log.WithValues("metal3remediation", req.NamespacedName)
+
+	// Fetch the Metal3Remediation instance.
+	metal3Remediation := &infrav1.Metal3Remediation{}
+	if err := r.Client.Get(ctx, req.NamespacedName, metal3Remediation); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		remediationLog.Error(err, "unable to get metal3Remediation")
+		return ctrl.Result{}, err
+	}
+
+	helper, err := patch.NewHelper(metal3Remediation, r.Client)
+	if err != nil {
+		remediationLog.Error(err, "failed to init patch helper")
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the Remediation object and status after each reconciliation.
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
+		patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+
+		patchErr := helper.Patch(ctx, metal3Remediation, patchOpts...)
+		if patchErr != nil {
+			remediationLog.Error(patchErr, "failed to Patch metal3Remediation")
+			// trigger requeue!
+			rerr = patchErr
+		}
+	}()
+
+	// Fetch the Machine.
+	// TODO refactor, put this into manager?
```

Comment:
Maybe let's do this now?

Reply:
Moved into its own method, but not into the manager, because the return value is actually needed for initializing the manager